core/src/service/api-creator.ts

```diff
@@ -0,0 +1,230 @@
+/* eslint-disable @typescript-eslint/consistent-type-assertions */
+/* eslint-disable react-hooks/rules-of-hooks */
+import React from 'react';
+import { createStore } from '../cube';
+import { isEmpty } from 'lodash';
+import axios from 'axios';
+import { Key, pathToRegexp, compile } from 'path-to-regexp';
+import qs from 'query-string';
+import { getConfig } from '../config';
+
+const DEFAULT_PAGESIZE = 15;
+
+/**
+ * Fill in the actual value for the path with parameters by path-to-regexp
+ * @param path Paths that may contain parameters, such as /fdp/:id/detail, path can not include `?`
+ * @param params The incoming parameters may be query or params
+ * @returns
+ */
+const generatePath = (path: string, params?: Obj) => {
+  try {
+    const toPathRepeated = compile(path);
+    return toPathRepeated(params);
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error('path:', path, 'Error parsing url parameters');
+    throw error;
+  }
+};
+
+/**
+ * Use path to match the incoming parameters, extract query and params
+ * @param path Paths that may contain parameters, such as /fdp/:id/detail, path can not include `?`
+ * @param params The incoming parameters may be query or params
+ */
+const extractPathParams = (path: string, params?: Record<string, any>) => {
+  const keys: Key[] = [];
+  pathToRegexp(path, keys);
+  const pathParams = {} as Obj<string>;
+  const bodyOrQuery = { ...params };
+  if (keys.length > 0) {
+    keys.forEach(({ name }) => {
+      pathParams[name] = bodyOrQuery[name];
+      delete bodyOrQuery[name];
+    });
+  }
+  return {
+    pathParams,
+    bodyOrQuery,
+  };
+};
+
+interface $options {
+  isDownload?: boolean; // whether its download api
+  uploadFileKey?: string; // upload formData attribute
+  autoMessage?: boolean; // eject message automatically
+  successMsg?: string; // eject message when success to override default message
+  errorMsg?: string; // eject message when failed to override default message
+}
+type $headers = Record<string, string>;
+type $body = Record<string, any>;
+interface CallParams {
+  $options?: $options;
+  $headers?: $headers;
+  $body?: $body;
+}
+
+type Merge<A, B> = { [K in keyof A]: K extends keyof B ? B[K] : A[K] } & B extends infer O
+  ? { [K in keyof O]: O[K] }
+  : never;
+
+/**
+ * generate api request function by config
+ * @param apiConfig
+ * @returns callable api function
+ */
+export const genRequest = function <T extends FN>(apiConfig: APIConfig) {
+  const { api, headers } = apiConfig;
+  let [method, path] = api.split('@');
+  if (!path) {
+    path = method;
+    method = 'get';
+  }
+
+  return (params: CallParams & Merge<Parameters<T>[0], {}>) => {
```

Comment:
no need declare Merge 
`params: CallParams & Parameters<T>[0] extends never ? {} : Parameters<T>[0]`

Reply:
well, Merge is used to extract the inner property out, so we can see the structure directly. 
![image](https://user-images.githubusercontent.com/3955437/123394947-4278e600-d5d2-11eb-82f5-f904f78e9655.png)